src/segment-loader.js

```diff
@@ -0,0 +1,400 @@
+     */
+    checkBuffer_: function(buffered, playlist, currentTime, timestampOffset) {
+      var currentBuffered = findRange(buffered, currentTime),
+          bufferedTime, currentBufferedEnd, discontinuity, segment, mediaIndex;
+
+      if (!playlist.segments.length) {
+        return;
+      }
+
+      timestampOffset = timestampOffset || 0;
+      if (currentBuffered.length === 0) {
+        // find the segment containing currentTime
+        mediaIndex = getMediaIndexForTime(playlist, currentTime, timestampOffset);
+      } else {
+        // find the segment adjacent to the end of the current
+        // buffered region
+        currentBufferedEnd = currentBuffered.end(0);
+        bufferedTime = Math.max(0, currentBufferedEnd - currentTime);
+
+        // if there is plenty of content buffered, relax for awhile
+        if (bufferedTime >= videojs.Hls.GOAL_BUFFER_LENGTH) {
+          return null;
+        }
+        mediaIndex = getMediaIndexForTime(playlist, currentBufferedEnd, timestampOffset);
+      }
+      segment = playlist.segments[mediaIndex];
+
+      // if the timestampOffset on the SourceBuffer is different from
+      // the timestampOffset for the closest discontinuity before the
+      // segment, update the timestampOffset:
+      // - find the closest discontinuity before the appended segment
+      // - set the SourceBuffer's timestampOffset to the estimated
+      //   start time of the first discontinuous segment
+      discontinuity = playlist.discontinuityStarts.slice().reverse()
+        .find(function(i) {
+          return i <= mediaIndex;
+        });
+      if (discontinuity) {
+        timestampOffset += duration(playlist, playlist.mediaSequence + mediaIndex);
+      }
+
+      return {
+        // resolve the segment URL relative to the playlist
+        uri: segment.resolvedUri,
+        // the segment's mediaIndex at the time it was requested
+        mediaIndex: mediaIndex,
+        // the segment's playlist
+        playlist: playlist,
+        // unencrypted bytes of the segment
+        bytes: null,
+        // when a key is defined for this segment, the encrypted bytes
+        encryptedBytes: null,
+        // optionally, the decrypter that is unencrypting the segment
+        decrypter: null,
+        // the state of the buffer before a segment is appended will be
+        // stored here so that the actual segment duration can be
+        // determined after it has been appended
+        buffered: null,
+        // The target timestampOffset for this segment when we append it
+        // to the source buffer
+        timestampOffset: timestampOffset
+      };
+    },
+
+    abort_: function() {
+      if (this.xhr_) {
+        // Prevent error handler from running.
+        this.xhr_.onreadystatechange = null;
+        this.xhr_.abort();
+        this.xhr_ = null;
+      }
+
+      // clear out the segment being processed
+      this.pendingSegment_ = null;
+    },
+
+    fillBuffer_: function() {
```

Comment:
If we are going through the effort to pull these functions into their own file, maybe we should consider giving `fillBuffer`/`drainBuffer` better names?